services/product.js

```
@@ -0,0 +1,171 @@
+
+    await saveProducts()
+}
+
+async function getByScrumMasterName(scrumMaster) {
+    const lowerScrumMasterNameNeeded = scrumMaster.toLowerCase()
+    let productsFound
+    if (scrumMaster) {
+        productsFound = (await getAll()).filter((product) => product.scrumMasterName.toLowerCase() === lowerScrumMasterNameNeeded)
+    } else {
+        productsFound = products.slice()
+    }
+   
+    if (!productsFound) {
+        throw new Error ("Product was not found!");
+    } else {
+        return productsFound;
+    }
+}
+
+async function getByDeveloperName(developer) {
+    const lowerDeveloperNameNeeded = developer.toLowerCase()
+    console.log("lowerDeveloperNameNeeded", lowerDeveloperNameNeeded)
+    let productsFound
+    if (developer) {
+        productsFound = (await getAll()).filter((product) => {
+            return product.Developers.reduce((acc, developer) => {
+                return acc || developer.toLowerCase() === lowerDeveloperNameNeeded
+            }, false)
+        })
+    } else {
+        productsFound = products.slice()
+    }
+
+    if (!productsFound) {
+        throw new Error ("Product was not found!");
+    } else {
+        return productsFound;
+    }
+}
+
+
+async function saveProducts() {
+    await fs.writeFile("data/products.json", JSON.stringify(products))
+}
+
+async function loadProducts() {
+    products = JSON.parse(await fs.readFile("data/products.json", "utf-8"))
+}
+
+//-- Helper functions for generating product details --
+
+function generateUId() {
+    let pId = ""
+    // generate new product ID until it doesn't collide with previously generated product IDs
+    do {
+        pId = "P"+(Math.random().toString(36).substr(2,5));
+    } while (products.reduce( (acc, product) => acc || (product.productId === pId), false)); 
+    
+    return pId
+}
+
+function createProduct(productId = "", productName= "", ownerName= "", arrDevelopers = [], scrumMasterName = "", startDate = "", methodology = "") {
```

Two things here:

1. I think it would be good to pass the values as an object, and not individual arguments.
2. All the arguments have a default value, is that really what you want? If an empty string for for example `startDate` is not valid, then this function allows invalid products to be created. It might be better to throw an error when something is missing.

---

1. Done, but I have a question: This is a helper function.  If I pass values as an object, the keys used in the helper function will need to agree with the keys of the object given by the user function.  Would it be better to leave values as input?
2. Done, but see also comment in next suggested change